Sources/SwiftBuildSupport/SwiftBuildSystemMessageHandler.swift

```diff
@@ -0,0 +1,539 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+@_spi(SwiftPMInternal)
+import Basics
+import Foundation
+@_spi(SwiftPMInternal)
+import SPMBuildCore
+import enum TSCUtility.Diagnostics
+import SWBBuildService
+import SwiftBuild
+import protocol TSCBasic.OutputByteStream
+
+
+/// Handler for SwiftBuildMessage events sent by the SWBBuildOperation.
+public final class SwiftBuildSystemMessageHandler {
+    private let observabilityScope: ObservabilityScope
+    private let logLevel: Basics.Diagnostic.Severity
+    private var buildState: BuildState = .init()
+
+    let progressAnimation: ProgressAnimationProtocol
+    var serializedDiagnosticPathsByTargetID: [Int: [Basics.AbsolutePath]] = [:]
+    var serializedDiagnosticPathsByTargetName: [String: [Basics.AbsolutePath]] {
```

Comment:
This is going to end up mixing diagnostics from potentially completely unrelated targets.

I see the existing BuildResult structure is forcing you into providing this, and it's only used by the Migrate command that was added recently. That should be refactored to use a globally unique identifier for the target instead (`ResolvedModule.ID`?), but that can be done separately from this PR.

For now I would add a FIXME comment that effectively repeats what I said above, though.